### PR TITLE
planner: check clustered index don't need double read

### PR DESCRIPTION
**Pull request by lzmhhh123**
### What problem does this PR solve?

Problem Summary: as the title says.

### What is changed and how it works?

What's Changed: check clustered index don't need double read

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM

### Release note 

- none

### CODE REVIEW

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -2859,6 +2859,15 @@ func (b *PlanBuilder) buildDataSource(ctx context.Context, tn *ast.TableName, as
 	ds.names = names
 	ds.setPreferredStoreType(b.TableHints())
 
+	// Init commonHandleCols and commonHandleLens for data source.
+	if tableInfo.IsCommonHandle {
+		for _, idx := range tableInfo.Indices {
+			if idx.Primary {
```
Comment (coocood) [inline planner/core/logical_plan_builder.go:2865]: There is a function `tables.FindPrimaryIndex` can be used.

File: planner/core/find_best_task.go [hunk 2]
```diff
@@ -700,17 +699,19 @@ func (ds *DataSource) buildIndexMergeTableScan(prop *property.PhysicalProperty,
...
+		if tblInfo.PKIsHandle && mysql.HasPriKeyFlag(col.RetType.Flag) {
 			continue
 		}
 		if col.ID == model.ExtraHandleID {
 			continue
 		}
 		isIndexColumn := false
-		for i, indexCol := range indexColumns {
-			isFullLen := idxColLens[i] == types.UnspecifiedLength || idxColLens[i] == col.RetType.Flen
+		for i, indexCol := range indexCols {
```
Comment (coocood) [inline planner/core/find_best_task.go:713]: We can extract this for loop, and call it twice with `indexCols` and `commonHandleCols`.
So we don't need to append slices.

File: planner/core/find_best_task.go [hunk 3]
```diff
@@ -814,52 +815,50 @@ func (is *PhysicalIndexScan) indexScanRowSize(idx *model.IndexInfo, ds *DataSour
...
+					RetType:  types.NewFieldType(mysql.TypeLonglong),
+					ID:       model.ExtraHandleID,
+					UniqueID: is.ctx.GetSessionVars().AllocPlanColumnID(),
+				})
+			}
+		}
+	} else {
+		if len(is.IdxCols) < len(is.Columns) {
+			for i := len(is.IdxCols); i < len(idxExprCols); i++ {
+				indexCols = append(indexCols, idxExprCols[i])
```
Comment (coocood) [inline planner/core/find_best_task.go:857]: Need to check if `idxExprCols[i] != nil`?

Comment (lzmhhh123) [reply planner/core/find_best_task.go:857]: `idxExprCols` here for common handle never be `nil`.

**Pull request merged**